sdk/identity/azure_identity/src/managed_identity_credential.rs

```diff
@@ -0,0 +1,498 @@
+            expected_source,
+            actual_source
+        );
+        let expires_on = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+            + 3600;
+        let mock_client = MockHttpClient::new(move |actual| {
+            {
+                let expected = model_request.clone();
+                let response_format = response_format.clone();
+                async move {
+                    assert_eq!(expected.method(), actual.method());
+
+                    let mut actual_params: Vec<_> =
+                        actual.url().query_pairs().into_owned().collect();
+                    actual_params.sort();
+                    let mut expected_params: Vec<_> =
+                        expected.url().query_pairs().into_owned().collect();
+                    expected_params.sort();
+                    assert_eq!(expected_params, actual_params);
+
+                    let mut actual_url = actual.url().clone();
+                    actual_url.set_query(None);
+                    let mut expected_url = expected.url().clone();
+                    expected_url.set_query(None);
+                    assert_eq!(actual_url, expected_url);
+
+                    // allow additional headers in the actual request so changing
+                    // the underlying client in the future won't break tests
+                    expected.headers().iter().for_each(|(k, v)| {
+                        assert_eq!(actual.headers().get_str(k).unwrap(), v.as_str())
+                    });
+
+                    Ok(Response::from_bytes(
+                        StatusCode::Ok,
+                        Headers::default(),
+                        Bytes::from(response_format.replacen(
+                            EXPIRES_ON,
+                            &expires_on.to_string(),
+                            1,
+                        )),
+                    ))
+                }
+            }
+            .boxed()
+        });
+        let mut options = options.unwrap_or_default();
+        options.credential_options = TokenCredentialOptions {
+            http_client: Arc::new(mock_client),
+            ..Default::default()
+        };
+        let cred = ManagedIdentityCredential::new(Some(options)).expect("credential");
+        let token = cred.get_token(LIVE_TEST_SCOPES).await.expect("token");
+        assert_eq!(token.expires_on.unix_timestamp(), expires_on as i64);
+        assert_eq!(token.token.secret(), "*");
+    }
+
+    fn run_unsupported_source_test(
+        env_vars: HashMap<&str, &str>,
+        expected_source: ManagedIdentitySource,
+    ) {
+        let _guard = EnvVarGuard::new(&env_vars);
+        let actual_source = get_source();
+        assert!(
+            std::mem::discriminant(&actual_source) == std::mem::discriminant(&expected_source),
+            "Expected {:?}, got {:?}",
+            expected_source,
+            actual_source
+        );
```

Comment:
```suggestion
        assert_eq!(std::mem::discriminant(&actual_source), std::mem::discriminant(&expected_source));
```

You can save yourself time by not putting all these "expected a, got b". That's not idiomatic in rust. The default error message is most often good enough. I only use the custom message when the default may not be as helpful.